kitsune/sumo/i18n.py

```diff
@@ -0,0 +1,138 @@
+    """
+    language, slash, rest_of_path = path.lstrip("/").partition("/")
+    return (language, f"{slash}{rest_of_path}")
+
+
+def normalize_language(language):
+    """
+    Given a language code, returns the language code supported by SUMO in the
+    proper case, for example "eN-us" --> "en-US" or "sC" --> "it", or None if
+    SUMO doesn't support the language code.
+    """
+    if not language:
+        return None
+    lc_language = language.lower()
+    return settings.LANGUAGE_URL_MAP.get(lc_language) or FALLBACK_LANGUAGE_URL_MAP.get(lc_language)
+
+
+def normalize_path(path, force_language=False):
+    """
+    Normalizes the language code, if any, that starts the path. If "force_language" is
+    given, it will replace any existing language code in the path, or prefix the path
+    if the path didn't already start with a language code.
+
+    Example when "force_language=None":
+        Requested  --> Normalized
+            /en-us --> /en-US  (normalization of case)
+            /fr-ca --> /fr     (supported variant)
+            /en-gb --> /en-US  (supported variant)
+            /sc    --> /it     (explicit fallback)
+            /ak    --> /en-US  (explicit fallback)
+
+    Example when "force_language=de":
+        Requested    --> Normalized
+            /path    --> /de/path  (path prefixed)
+            /it/path --> /de/path  (language code replaced)
+    """
+    # Use Django's standard way to check if the path starts with a supported language.
+    language_from_path = translation.get_language_from_path(path)
+
+    # Even if Django found that the path didn't start with a supported language, the path
+    # may start with one for which SUMO defines a supported fallback language, for example
+    # "sc" --> "it".
+    if not language_from_path and (mo := LANGUAGES_WITH_FALLBACKS_REGEX.match(path)):
+        language_from_path = mo.group("language")
+
+    normalized_language_from_path = normalize_language(language_from_path)
+
+    if normalized_language_from_path:
+        actual_language_from_path, remaining_path = split_into_language_and_path(path)
+        if force_language:
+            return f"/{force_language}{remaining_path}"
+        elif actual_language_from_path != normalized_language_from_path:
+            return f"/{normalized_language_from_path}{remaining_path}"
+    elif force_language:
+        return f"/{force_language}{path}"
+
+    return path
+
+
+def get_language_from_request(request, check_path=False, check_user=False):
+    """
+    A wrapper around Django's "get_language_from_request" that provides SUMO-specific
+    customizations:
+        1) Check for a logged-in user's locale in their profile
+        2) Check the user's session for a locale
+        3) Always return a language code normalized for SUMO
+    """
+    language = None
+
+    if check_path:
+        language = translation.get_language_from_path(request.path_info)
+
+    if check_user and not language:
```

Comment:
Do we actually need the check for `language` if the `check_user` is set? Or reversing this, do we need the `check_user` if we actually check for `language`? One of the two feels redundant.

Reply:
super nit: maybe this is a nice case for a `match` statement

Reply:
> super nit: maybe this is a nice case for a `match` statement

I don't think it would work well in this case, due to the fact that we flow through to the next step if a language isn't found in the current step.